### PR TITLE
Implement profile and level privacy settings

### LINKED ISSUE #568: Privacy Settings
The following privacy settings should be added to the website for users to configure:

- [x] Blocking other users
- [ ] Enabling/disabling profile comments
- [ ] Enabling/disabling level comments

There are also two in-game privacy settings we could choose to enable or disable as well. The options are to allow your levels and profile to be seen by everyone, only logged in users, or no one on the website. For public beta I'm not too concerned about these. Regardless of the user's choice, moderators should always be able to see the levels and profile on the website

### PR DESCRIPTION
**Pull request by sudokoko**
This PR implements a new settings area allowing you to set your privacy settings for your profile/levels. This recreates the functionality that LBP.me had, however these options would be set in game rather than through the website when LBP.me was a thing.

image

From the new settings page, you're able to **enable/disable your profile comments** (does not allow you to override an active UserDisableComments case), as well as **set your profile/level visibility settings**.

You're able to use the following three options for profile/level privacy;
* Public - shows your profile or levels to everyone
* Signed-In Only - only shows your profile or levels to users who are signed into the site
* In-Game Only - only shows your profile or levels to users who are in-game.

<details>
<summary>Further images</summary>
<br>

image
image
image
image

</details>

Closes #568.

### CODE REVIEW

File: ProjectLighthouse.Servers.Website/Pages/Partials/CommentsPartial.cshtml [hunk 1]
```diff
@@ -1,5 +1,4 @@
 ﻿@using System.Web
-@using System.IO
```
Comment (Zaprit) [inline ProjectLighthouse.Servers.Website/Pages/Partials/CommentsPartial.cshtml:2]: why did this even have system.io imported anyway????

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml [hunk 2]
```diff
@@ -106,172 +113,199 @@
...
+        </div>
+        @if (isMobile)
+        {
+            <br/>
+        }    
     }
 </div>
 
-<div class="ui grid">
-    @{
```
Comment (Zaprit) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml:139]: probably out the scope of this PR but UI changes should NOT be done on the serverside, there's a whole language for this, it's called CSS (this is more as a general thing for lighthouse as a whole, but this here is a really good example of why)

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 3]
```diff
@@ -18,44 +18,82 @@ public class UserPage : BaseLayout
...
+                    break;
+                }
+                case PrivacyType.All:
+                {
+                    this.LevelsPrivate = false;
+                    break;
+                }
+                default:
+                {
+                    this.LevelsPrivate = false;
```
Comment (Zaprit) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:94]: This can be consolidated to save an if
```suggestion
        if (this.User == null || !this.User.IsModerator)
        {
            // Determine if user can view profile according to profileUser's privacy settings
            switch (this.ProfileUser.ProfileVisibility)
            {
                case PrivacyType.Game:
                {
                    if (this.ProfileUser != this.User) this.ProfilePrivate = true;
                    break;
                }
                case PrivacyType.PSN:
                {
                    if (this.User == null) this.ProfilePrivate = true;
                    break;
                }
                case PrivacyType.All:
                {
                    this.ProfilePrivate = false;
                    break;
                }
                default:
                {
                    this.ProfilePrivate = false;
                    break;
                }
            }

            // Determine if user can view levels according to profileUser's privacy settings
            switch (this.ProfileUser.LevelVisibility)
            {
                case PrivacyType.Game:
                {
                    if (this.ProfileUser != this.User) this.LevelsPrivate = true;
                    break;
                }
                case PrivacyType.PSN:
                {
                    if (this.User == null) this.LevelsPrivate = true;
                    break;
                }
                case PrivacyType.All:
                {
                    this.LevelsPrivate = false;
                    break;
                }
                default:
                {
                    this.LevelsPrivate = false;
                    break;
                }
            }
        }
```

File: ProjectLighthouse.Servers.Website/Pages/UsersPage.cshtml.cs [hunk 4]
```diff
@@ -37,8 +37,9 @@ public async Task<IActionResult> OnGet([FromRoute] int pageNumber, [FromQuery] s
 
         if (this.PageNumber < 0 || this.PageNumber >= this.PageAmount) return this.Redirect($"/users/{Math.Clamp(this.PageNumber, 0, this.PageAmount - 1)}");
 
-        this.Users = await this.Database.Users.Where(u => u.PermissionLevel != PermissionLevel.Banned && u.Username.Contains(this.SearchValue))
-            .Where(u => u.ProfileVisibility == PrivacyType.All) // TODO: change check for when user is logged in
+        this.Users = await this.Database.Users
```
Comment (Zaprit) [inline ProjectLighthouse.Servers.Website/Pages/UsersPage.cshtml.cs:40]: Should people with private profiles be included in this list?

**Review by Zaprit — CHANGES_REQUESTED** [review]
See suggestions

File: ProjectLighthouse.Servers.Website/Pages/Partials/CommentsPartial.cshtml (hunk 1 above)
Comment (sudokoko) [reply ProjectLighthouse.Servers.Website/Pages/Partials/CommentsPartial.cshtml:2]: Not sure. Rider removed this on it's own

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml (hunk 2 above)
Comment (sudokoko) [reply ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml:139]: I believe slendy is working on a solution for this, not sure

File: ProjectLighthouse.Servers.Website/Pages/UsersPage.cshtml.cs (hunk 4 above)
Comment (sudokoko) [reply ProjectLighthouse.Servers.Website/Pages/UsersPage.cshtml.cs:40]: This was done on purpose - since you should still be able to search for a user, as the information that would be shown there is the same thing shown on the user page anyways regardless of privacy settings.

**Review by sudokoko — COMMENTED** [review]
Responses to @Zaprit

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 5]
```diff
@@ -18,44 +19,81 @@ public class UserPage : BaseLayout
...
+    { }
 
     public async Task<IActionResult> OnGet([FromRoute] int userId)
     {
         this.ProfileUser = await this.Database.Users.FirstOrDefaultAsync(u => u.UserId == userId);
         if (this.ProfileUser == null) return this.NotFound();
 
         // Determine if user can view profile according to profileUser's privacy settings
-        if (this.User == null || !this.User.IsAdmin)
+        if (this.User == null || !this.User.IsModerator)
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:48]: I don't know if this was intentional or not but on the UserPage you let mods bypass privacy settings but on the SlotPage you only let admins bypass the privacy settings

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 6]
```diff
@@ -18,44 +19,81 @@ public class UserPage : BaseLayout
...
                 {
-                    if (this.ProfileUser != this.User) return this.NotFound();
+                    this.ProfilePrivate = false;
+                    break;
+                }
+                default:
+                {
+                    this.ProfilePrivate = false;
+                    break;
+                }
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:71]: This can be simplified into just the default block

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 7]
```diff
@@ -18,44 +19,81 @@ public class UserPage : BaseLayout
...
+                case PrivacyType.All:
+                {
+                    this.SlotsPrivate = false;
+                    break;
+                }
+                default:
+                {
+                    this.SlotsPrivate = false;
                     break;
                 }
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:96]: Simplify
```suggestion
                default:
                {
                    this.SlotsPrivate = false;
                    break;
                }
```
Or if rider complains put both labels above the same block

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 8]
```diff
@@ -116,7 +157,8 @@ public async Task<IActionResult> OnGet([FromRoute] int userId)
 
         foreach (KeyValuePair<CommentEntity, RatedCommentEntity?> kvp in this.Comments)
         {
-            RatedCommentEntity? reaction = await this.Database.RatedComments.Where(r => r.CommentId == kvp.Key.CommentId)
+            RatedCommentEntity? reaction = await this.Database.RatedComments
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:160]: I understand that you prefer this style of indentation and I'm open to having a conversation about changing the entire project to this style, but for now, instead of changing them one by one we should just use the old style and then at a later date we can reformat the entire project

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                    <select class="ui fluid dropdown" type="text" name="profilePrivacyLevel" id="profilePrivacyLevel">
+                        <option value="public" @(Model.ProfileUser.ProfileVisibility == PrivacyType.All ? "selected" : "")>
+                            Share your profile with everyone!
+                        </option>
+                        <option value="signedInOnly" @(Model.ProfileUser.ProfileVisibility == PrivacyType.PSN ? "selected" : "")>
+                            Only share your profile with users who are signed into the website or playing in-game.
+                        </option>
+                        <option value="inGameOnly" @(Model.ProfileUser.ProfileVisibility == PrivacyType.Game ? "selected" : "")>
+                            Only share your profile with users who are playing in-game.
+                        </option>
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml:34]: Since these are all the same except for the value or selected tag, you could try and simplify this by iterating over all possible values of PrivacyType (or keeping a static cached reference of all types) and then for each option the value would just be the ToString() of the type and the selected would be applied if the users profile visibility equals that type

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+                    </select>
+                </div>
+                <div class="field">
+                    <label style="text-align: left" for="profileCommentsEnabled">
+                        Comments
+                        @if (Model.CommentsDisabledByModerator)
+                        {
+                            <small style="color: red">Locked by a moderator</small>
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml:42]: Does Fomantic have classes for formatting text so we don't have to use inline styles?

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+    public UserPrivacyPage(DatabaseContext database) : base(database)
+    { }
+
+    public async Task<IActionResult> OnGet([FromRoute] int userId)
+    {
+        this.ProfileUser = await this.Database.Users.FirstOrDefaultAsync(u => u.UserId == userId);
+        if (this.ProfileUser == null) return this.NotFound();
+
+        if (this.User == null) return this.Redirect("~/user/" + userId);
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs:27]: this should probably redirect to /login instead

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        if (this.User != this.ProfileUser) return this.Redirect("~/user/" + userId);
+
+        this.BlockedUsers = await this.Database.BlockedProfiles.Where(b => b.UserId == this.ProfileUser.UserId)
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs:30]: I haven't tested this myself yet but does this work without using `.Include(b => b.BlockedUser)` before the select?

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+        this.CommentsDisabledByModerator = await this.Database.Cases.Where(c => c.AffectedId == this.ProfileUser.UserId)
+            .Where(c => c.Type == CaseType.UserDisableComments)
+            .Where(c => c.DismissedAt == null)
+            .AnyAsync();
+
+        return this.Page();
+    }
+
+    public async Task<IActionResult> OnPost([FromRoute] int userId, [FromForm] string profilePrivacyLevel, [FromForm] string profileCommentsEnabled, [FromForm] string slotPrivacyLevel)
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs:42]: since profileCommentsEnabled is being parsed into true or false anyways does making it a bool work so then ASP.NET will try to implicitly cast it

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+            _ => this.ProfileUser.ProfileVisibility,
+        };
+
+        this.ProfileUser.LevelVisibility = slotPrivacyLevel switch
+        {
+            "public" => PrivacyType.All,
+            "signedInOnly" => PrivacyType.PSN,
+            "inGameOnly" => PrivacyType.Game,
+            _ => this.ProfileUser.LevelVisibility,
+        };
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs:83]: Since you use the same function twice you should extract it at least into a local function or a private instance function

File: ProjectLighthouse.Servers.Website/Pages/UsersPage.cshtml.cs [hunk 15]
```diff
@@ -37,8 +37,9 @@ public async Task<IActionResult> OnGet([FromRoute] int pageNumber, [FromQuery] s
...
+            .Where(u => u.Username.Contains(this.SearchValue))
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UsersPage.cshtml.cs:41]: The same thing about the chaining style, we should either reformat the entire project or keep using the old style. We shouldn't be mixing

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+                "true" => true,
+                "false" => false,
+                _ => this.ProfileUser.CommentsEnabled,
+            };
+        }
+        else
+        {
+            this.ProfileUser.CommentsEnabled = false;
+        }
+
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs:68]: Why is this necessary if users are already prevented from changing whether their comments are enabled

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        return this.Page();
+    }
+
+    public async Task<IActionResult> OnPost([FromRoute] int userId, [FromForm] string profilePrivacyLevel, [FromForm] string profileCommentsEnabled, [FromForm] string slotPrivacyLevel)
+    {
+        this.ProfileUser = await this.Database.Users.FirstOrDefaultAsync(u => u.UserId == userId);
+        if (this.ProfileUser == null) return this.NotFound();
+
+        if (this.User == null) return this.Redirect("~/user/" + userId);
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs:47]: Same thing here, should probably be redirected to /login

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 18]
```diff
@@ -128,7 +170,13 @@ public async Task<IActionResult> OnGet([FromRoute] int userId)
             .AnyAsync();
 
         this.IsProfileUserBlocked = await this.Database.IsUserBlockedBy(this.ProfileUser.UserId, this.User.UserId);
-        
+
+        this.CommentsDisabledByModerator = await this.Database.Cases
+            .Where(c => c.AffectedId == this.ProfileUser.UserId)
+            .Where(c => c.Type == CaseType.UserDisableComments)
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:176]: linq style

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 19]
```diff
@@ -18,44 +19,81 @@ public class UserPage : BaseLayout
 
     public bool CommentsEnabled;
 
-    public bool IsProfileUserHearted;
+    public List<SlotEntity>? HeartedSlots;
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:22]: Is this meant to be in alphabetical order? If it is, then you should have the same types grouped together then the groups should be sorted alphabetically.

File: ProjectLighthouse.Servers.Website/Pages/SlotPage.cshtml.cs [hunk 20]
```diff
@@ -37,22 +39,28 @@ public async Task<IActionResult> OnGet([FromRoute] int id)
         // Determine if user can view slot according to creator's privacy settings
         if (this.User == null || !this.User.IsAdmin)
         {
-            switch (slot.Creator.ProfileVisibility)
+            switch (slot.Creator.LevelVisibility)
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/SlotPage.cshtml.cs:42]: This entire calculation of whether a profile should be private or not could be refactored into its own function that takes a privacy level and a user and outputs a bool

**Review by Slendy — CHANGES_REQUESTED** [review]
Everything seems functional, most concerns are clarifications, about styling, or refactoring repeated code

File: ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs (hunk 12 above)
Comment (sudokoko) [reply ProjectLighthouse.Servers.Website/Pages/UserPrivacyPage.cshtml.cs:30]: Yes

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 21]
```diff
@@ -29,35 +31,25 @@ public class UserPage : BaseLayout
...
-                case PrivacyType.All: break;
-                default: throw new ArgumentOutOfRangeException();
-            }
-        }
+        this.CanAccessProfile = this.ProfileUser.ProfileVisibility.CanAccess(
+            this.User != null, 
+            this.ProfileUser == this.User || this.User != null && this.User.IsModerator);
+        this.CanAccessSlots = this.ProfileUser.LevelVisibility.CanAccess(
+            this.User != null, 
+            this.ProfileUser == this.User || this.User != null && this.User.IsModerator);
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:52]: You could extract the authenticated and owner bools so you don't have to calculate them twice. Something like this: <sub>(please don't commit this directly I'm writing this in a browser)</sub>
```suggestion
        bool isLoggedIn = this.User != null;
        bool isOwner = this.ProfileUser == this.User || this.User != null && this.User.IsModerator;
        this.CanAccessProfile = this.ProfileUser.ProfileVisibility.CanAccess(isLoggedIn, isOwner);
        this.CanAccessSlots = this.ProfileUser.LevelVisibility.CanAccess(isLoggedIn, isOwner);
```

File: ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs [hunk 22]
```diff
@@ -29,35 +31,25 @@ public class UserPage : BaseLayout
     public List<SlotEntity>? QueuedSlots;
 
     public UserEntity? ProfileUser;
+
+    public bool CanAccessProfile;
+    public bool CanAccessSlots;
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/UserPage.cshtml.cs:36]: This might infer context better if it were named CanViewProfile and CanViewSlots

File: ProjectLighthouse.Servers.Website/Pages/SlotPage.cshtml.cs [hunk 23]
```diff
@@ -22,6 +22,8 @@ public class SlotPage : BaseLayout
     public bool CommentsEnabled;
     public readonly bool ReviewsEnabled = ServerConfiguration.Instance.UserGeneratedContentLimits.LevelReviewsEnabled;
 
+    public bool CanAccessSlot;
```
Comment (Slendy) [inline ProjectLighthouse.Servers.Website/Pages/SlotPage.cshtml.cs:25]: Same suggestion, should this be named CanViewSlot?

**Review by Slendy — CHANGES_REQUESTED** [review]
just some nitpicks about naming and duplicate code

**Review by Slendy — APPROVED** [review]

File: ProjectLighthouse/Types/Users/PrivacyType.cs [hunk 24]
```diff
@@ -21,6 +21,17 @@ public enum PrivacyType
 
 public static class PrivacyTypeExtensions
 {
+    public static string ToReadableString(this PrivacyType type, string area)
```
Comment (Zaprit) [inline ProjectLighthouse/Types/Users/PrivacyType.cs:24]: Should this be translated

**Review by Zaprit — CHANGES_REQUESTED** [review]
See comment

File: ProjectLighthouse/Types/Users/PrivacyType.cs (hunk 24 above)
Comment (sudokoko) [reply ProjectLighthouse/Types/Users/PrivacyType.cs:24]: I have no problem adding translations for it. One moment

**Review by Zaprit — APPROVED** [review]
LGTM
